cache/remotecache/s3/global_importer.go

```diff
@@ -0,0 +1,291 @@
+
+func (f importerFunc) Resolve(ctx context.Context, desc ocispecs.Descriptor, id string, w worker.Worker) (solver.CacheManager, error) {
+	return f(ctx, desc, id, w)
+}
+
+type globalImporter struct {
+	ctx      context.Context
+	id       string
+	s3Client *s3Client
+	config   Config
+	worker   worker.Worker
+
+	remotesLock sync.RWMutex
+	getByID     map[string]item
+}
+
+type item struct {
+	createdAt time.Time
+	result    *solver.Remote
+}
+
+type globalKeyImporter struct {
+	*globalImporter
+}
+
+func (gki *globalKeyImporter) AddLink(id string, link solver.CacheInfoLink, target string) error {
+	return nil // the importer is read only
+}
+
+func (gki *globalKeyImporter) AddResult(id string, res solver.CacheResult) error {
+	return nil // the importer is read only
+}
+
+func (gki *globalKeyImporter) Exists(id string) bool {
+	key := gki.s3Client.prefix + manifests + id
+	createdAt, _ := gki.s3Client.exists(gki.ctx, key)
+	return createdAt != nil
+}
+
+func (gki *globalKeyImporter) HasLink(id string, link solver.CacheInfoLink, target string) bool {
+	return false // TODO if needed
+}
+
+func (gki *globalKeyImporter) Load(id string, resultID string) (solver.CacheResult, error) {
+	gki.remotesLock.RLock()
+	item, found := gki.getByID[resultID]
+	gki.remotesLock.RUnlock()
+	if !found {
+		return solver.CacheResult{}, errors.WithStack(solver.ErrNotFound)
+	}
+	return solver.CacheResult{
+		ID:        remoteID(item.result),
+		CreatedAt: item.createdAt,
+	}, nil
+}
+
+func (gki *globalKeyImporter) Release(resultID string) error {
+	return nil // the importer is read only and this should not happen
+}
+
+func (gki *globalKeyImporter) Walk(fn func(id string) error) error {
+	return nil // explicitly not listing all possible manifests
+}
+
+func (gki *globalKeyImporter) WalkIDsByResult(resultID string, fn func(string) error) error {
+	return nil // TODO if needed
+}
+
+func (gki *globalKeyImporter) WalkBacklinks(id string, fn func(id string, link solver.CacheInfoLink) error) error {
+	if !isCachableKey(id) {
+		return nil
+	}
+	prefix := gki.s3Client.prefix + backlinks + id + separator
+	return gki.s3Client.ListAllOjectsV2Prefix(gki.ctx, prefix, func(out *s3.ListObjectsV2Output) (cont bool, err error) {
+		for _, obj := range out.Contents {
+			parts := strings.Split(strings.TrimPrefix(*obj.Key, prefix), separator)
+			if len(parts) != 4 {
+				// TODO: log something, but how
+				continue
+			}
+			part1, err := strconv.Atoi(parts[1])
+			if err != nil {
+				// TODO: log something, but how
+				continue
+			}
+			// `ID::dgst::inputIdx::selector::parent.ID`
+			// backlink from parent to id
+			fn(parts[3], solver.CacheInfoLink{
+				Digest:   digest.Digest(parts[0]),
+				Input:    solver.Index(part1),
+				Selector: digest.Digest(parts[2]),
+			})
+
+		}
+		return true, nil
+	})
+}
+
+func (gki *globalKeyImporter) WalkLinks(id string, link solver.CacheInfoLink, fn func(id string) error) error {
```

Comment:
Question here, could it be a good/better idea to store the ID in the file, instead of adding it as a suffix, and simply open the file without listing ?

To what degree here do you think there could be more than one file ? To me it feels like there should generally only be one, with correct handling of the `link`, but I wasn't certain and since this is called `WalkLinks` and other implemeters were lists, I made it a list.

Reply:
Trying this to see.

Reply:
Okay, making the links plain files, seems to make the things faster, but this is only a feeling, I'm trying to find out how to measure here; trying to see if I can implement some OTEL tracing to get a local call graph (#4443)